[multistage] [feature] Add a query option to pass some v1 limit

61yao opened the pull request:
Add a query option to pass in leaf node limit for v2 engine,  numGroupsLimit,  _maxInitialResultHolderCapacity, and groupByTrimThreshold.

This is needed because we are testing some tables that exceed the limit and noticed incorrect results.

Unfortunately, I am not able to test this because we are not comparing the row by row in integration test so I added a log to verify.

walterddr commented:
thanks for creating this. i think there are several more configurations in V1 we have to check
1. is DEFAULT_LIMIT which is 10 --> we changed it to 10Million for V2; but we can make it configurable
2. GROUP LIMIT REACHED early termination --> we should always set this to false.
3. (there might be more)

61yao commented:
> thanks for creating this. i think there are several more configurations in V1 we have to check
> 
> 1. is DEFAULT_LIMIT which is 10 --> we changed it to 10Million for V2; but we can make it configurable
> 2. GROUP LIMIT REACHED early termination --> we should always set this to false.
> 3. (there might be more)

Updated. PTAL.

walterddr commented on pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java:
```diff
@@ -291,6 +291,21 @@ public static class QueryOptionKey {
+        public static final String MS_LEAF_LIMIT = "multiStageLeafLimit";
+
+        // Override numGroupLimit used by v1 engine.
+        public static final String NUM_GROUP_LIMIT = "numGroupLimit";
+
+        // Override maxInitialResultCap used by v1 engine.
+        public static final String MAX_INITIAL_RESULT_HOLDER_CAPACITY = "maxInitialResultCap";
+
+        // Override groupByTrimThreshold used by v1 engine.
+        public static final String GROUP_BY_TRIM_THRESHOLD = "groupByTrimThreshold";
```
@Jackie-Jiang ^ any idea whether we want to allow modify on all 3 or should we derive them from a basic setting?

walterddr replied:
@snleee any thoughts on this^

Jackie-Jiang replied:
We can make all of them configurable from the query option to be flexible

siddharthteotia replied:
I think we have typically played with `numGroupsLimit` vast majority of the times and for some of our production use cases have set it higher than default 100K to get more accurate results for high cardinality GROUP BY. 

I don't remember configuring the `resultHolderCapacity` for any use case.

We have configured `trimThreshold` in one case for performance reasons

siddharthteotia replied:
Also IIRC, the `trimThreshold` option when it was introduced earlier was specifically applied at server level execution during combine in `ConcurrentIndexedTable`. Later it was pushed down into the segment level as well afaik. So, will be good to double check the expectations / implications of setting this in the context of multi-stage depending on desired behavior.

walterddr replied:
sounds good. thanks for sharing your opinions

walterddr commented on pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/MultiStageEngineIntegrationTest.java:
```diff
@@ -99,6 +99,14 @@ public void testGeneratedQueries()
+  @Test
+  public void testQueryOptions()
+      throws Exception {
+    String pinotQuery = "SELECT * FROM mytable OPTION(multiStageLeafLimit=1)";
```
let's use `SET multistageLeafLimit = 1` as the `OPTION` usage is being deprecated.
also how come H2 and pinot return the same result? i thought specifically this option makes the leaf only return 1 row right?

Jackie-Jiang commented on pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java:
```diff
@@ -291,6 +291,21 @@ public static class QueryOptionKey {
+        // It has to be greater than 0.
+        // If it is invalid number format, or the number is smaller than or equal to zero, we use
+        // DEFAULT_LEAF_NODE_LIMIT.
+        public static final String MS_LEAF_LIMIT = "multiStageLeafLimit";
+
+        // Override numGroupLimit used by v1 engine.
+        public static final String NUM_GROUP_LIMIT = "numGroupLimit";
+
+        // Override maxInitialResultCap used by v1 engine.
+        public static final String MAX_INITIAL_RESULT_HOLDER_CAPACITY = "maxInitialResultCap";
+
+        // Override groupByTrimThreshold used by v1 engine.
+        public static final String GROUP_BY_TRIM_THRESHOLD = "groupByTrimThreshold";
```
Suggest keeping the naming convention:
```suggestion
        public static final String MULTI_STAGE_LEAF_LIMIT = "multiStageLeafLimit";
        public static final String NUM_GROUPS_LIMIT = "numGroupsLimit";
        public static final String MAX_INITIAL_RESULT_HOLDER_CAPACITY = "maxInitialResultHolderCapacity";
        public static final String GROUP_TRIM_THRESHOLD = "groupTrimThreshold";
```

61yao replied:
Done

siddharthteotia commented on pinot-core/src/main/java/org/apache/pinot/core/util/QueryOptionsUtils.java:
```diff
@@ -105,4 +105,56 @@ public static boolean isServerReturnFinalResult(Map<String, String> queryOptions
     return queryOptions.get(QueryOptionKey.ORDER_BY_ALGORITHM);
   }
+
+  public static int getMultiStageLeafLimit(Map<String, String> queryOptions, int defaultLeafLimit) {
```
Curious to know what will be the use of this option to configure leaf limit ?

The other existing options should be enough to control the GROUP BY output / accuracy at leaf layer ?

61yao replied:
If # of rows scanned exceeds the limit, we may return wrong result right? 

The principle here is set up like limit or cap should be configurable via things other than codes. 
Cluster config is also good but I guess right now, it would be good to start with an option in query and we eventually move some of them to cluster config if we don't need to change it query by query. 

I admit it is kinda hacky to set everything via query option.. but I don't know a better way to make stuff configurable

61yao commented on pinot-core/src/main/java/org/apache/pinot/core/util/QueryOptionsUtils.java:
```diff
@@ -105,4 +105,56 @@ public static boolean isServerReturnFinalResult(Map<String, String> queryOptions
     return queryOptions.get(QueryOptionKey.ORDER_BY_ALGORITHM);
   }
+
+  public static int getMultiStageLeafLimit(Map<String, String> queryOptions, int defaultLeafLimit) {
```
Done

walterddr commented on pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultiStageBrokerRequestHandler.java:
```diff
@@ -165,7 +165,8 @@ private BrokerResponseNative handleRequest(long requestId, String query,
     ResultTable queryResults;
     try {
-      queryResults = _queryDispatcher.submitAndReduce(requestId, queryPlan, _mailboxService, queryTimeoutMs);
+      queryResults = _queryDispatcher.submitAndReduce(requestId, queryPlan, _mailboxService, queryTimeoutMs,
+          sqlNodeAndOptions.getOptions());
```
nit: please use pinot intellij checkstyle so that it doesn't automatically reformat code like these changes

61yao replied:
Done

walterddr commented on pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java:
```diff
@@ -209,27 +209,39 @@ private void applyQueryOptions(QueryContext queryContext) {
       // Set maxInitialResultHolderCapacity
-      queryContext.setMaxInitialResultHolderCapacity(_maxInitialResultHolderCapacity);
-
+      Integer initResultCap = QueryOptionsUtils.getMaxInitResultCap(queryOptions);
+      if (initResultCap != null) {
+        queryContext.setMaxInitialResultHolderCapacity(initResultCap);
+      } else {
+        queryContext.setMaxInitialResultHolderCapacity(_maxInitialResultHolderCapacity);
+      }
```
nit: unboxing can be coupled with default.

```suggestion
      Integer initResultCap = QueryOptionsUtils.getMaxInitResultCap(queryOptions);
      int maxInitialResultHolderCapacity = initResultCap == null ? _maxInitialResultHolderCapacity : initResultCap;
      queryContext.setMaxInitialResultHolderCapacity(maxInitialResultHolderCapacity);
```

61yao replied:
Done

walterddr commented on pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/MultiStageEngineIntegrationTest.java:
```diff
@@ -99,6 +99,14 @@ public void testGeneratedQueries()
+  public void testQueryOptions()
+      throws Exception {
+    String pinotQuery = "SET multistageLeafLimit = 1; SELECT * FROM mytable;";
+    String h2Query = "SELECT * FROM mytable";
```
curious how did this test passed? don't you need to do `SELECT * FROM mytable LIMIT 1` for H2 query?

61yao replied:
Current test frame work only check H2 query contains the request for pinot query

Jackie-Jiang replied:
Hmm, this test cannot prove the option is taking effect. Can you also verify that the pinot query only returns one record?

walterddr replied:
great catch @Jackie-Jiang. i think we can directly assert 1 row instead of comparing against h2

61yao replied:
Added a test to compare # of rows. PTAL.

walterddr commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/service/QueryDispatcher.java:
```diff
@@ -89,7 +89,9 @@ public int submit(long requestId, QueryPlan queryPlan, long timeoutMs)
               .putMetadata(QueryConfig.KEY_OF_BROKER_REQUEST_ID, String.valueOf(requestId))
-              .putMetadata(QueryConfig.KEY_OF_BROKER_REQUEST_TIMEOUT_MS, String.valueOf(timeoutMs)).build());
+              .putMetadata(QueryConfig.KEY_OF_BROKER_REQUEST_TIMEOUT_MS, String.valueOf(timeoutMs))
+              .putAllMetadata(queryOptions).build());
```
should we only filter out queryOptions and attach only those that is listed in `QueryConfig.KEY_OF_*`?

61yao replied:
it is easier to pass in query option because we may want to pass a lot more things in the future

walterddr replied:
on a 2nd thought i agree we should allow passing everything. current V1 engine also doesn't limit the KEY options. we can add if we saw issues in the future

walterddr reviewed (commented):
lgtm overall. please kindly take a look at the comments

walterddr commented:
@61yao could we address the comments and rebase this PR?

Jackie-Jiang commented on pinot-common/src/main/java/org/apache/pinot/common/utils/config/QueryOptionsUtils.java:
```diff
@@ -166,4 +166,28 @@ public static boolean isServerReturnFinalResult(Map<String, String> queryOptions
+  }
+
+  @Nullable
+  public static Integer getNumGroupLimit(Map<String, String> queryOptions) {
```
Suggest keeping the name consistent for readability, same for other getters
```suggestion
  public static Integer getNumGroupsLimit(Map<String, String> queryOptions) {
```

Jackie-Jiang commented on pinot-common/src/main/java/org/apache/pinot/common/utils/config/QueryOptionsUtils.java:
```diff
@@ -166,4 +166,28 @@ public static boolean isServerReturnFinalResult(Map<String, String> queryOptions
+  }
+
+  @Nullable
+  public static Integer getMaxInitResultCap(Map<String, String> queryOptions) {
```
```suggestion
  public static Integer getMaxInitialResultHolderCapacity(Map<String, String> queryOptions) {
```

Jackie-Jiang commented on pinot-common/src/main/java/org/apache/pinot/common/utils/config/QueryOptionsUtils.java:
```diff
@@ -166,4 +166,28 @@ public static boolean isServerReturnFinalResult(Map<String, String> queryOptions
+  }
+
+  @Nullable
+  public static Integer getGroupByTrimThreshold(Map<String, String> queryOptions) {
```
```suggestion
  public static Integer getGroupTrimThreshold(Map<String, String> queryOptions) {
```

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java:
```diff
@@ -209,27 +209,36 @@ private void applyQueryOptions(QueryContext queryContext) {
       // Set maxInitialResultHolderCapacity
-      queryContext.setMaxInitialResultHolderCapacity(_maxInitialResultHolderCapacity);
-
+      Integer initResultCap = QueryOptionsUtils.getMaxInitResultCap(queryOptions);
```
Let's use the same way to set the arguments for readability, e.g. always do if check

Jackie-Jiang reviewed (approved):
Minor comments. @walterddr Please help review the v2 part change
